src/mpid/ch4/shm/xpmem/xpmem_impl.h

```diff
@@ -62,34 +64,48 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_seg_regist(int node_rank, size_t size,
         XPMEM_DBG_PRINT("register apid 0x%lx for node_rank %d, segid 0x%lx\n",
                         (uint64_t) segmap->apid, node_rank, (uint64_t) segmap->remote_segid);
     }
-    segmap->refcount++;
-
-    /* Create segment. Both offset and size must be page aligned. */
-    /* TODO: use pre-allocated buffer */
-    MPIR_CHKPMEM_MALLOC(seg, MPIDI_XPMEM_seg_t *, sizeof(MPIDI_XPMEM_seg_t),
-                        mpi_errno, "xpmem seg", MPL_MEM_SHM);
-    seg->offset = MPL_ROUND_DOWN_ALIGN((uint64_t) remote_vaddr,
-                                       (uint64_t) MPIDI_XPMEM_global.sys_page_sz);
-    offset_diff = (off_t) remote_vaddr - seg->offset;
-    seg->size = MPL_ROUND_UP_ALIGN(size + (size_t) offset_diff, MPIDI_XPMEM_global.sys_page_sz);
-    seg->segmap_ptr = segmap;
-
-    xpmem_addr.apid = segmap->apid;
-    xpmem_addr.offset = seg->offset;
-    seg->vaddr = xpmem_attach(xpmem_addr, seg->size, NULL);
-    /* virtual address or failure(-1) */
-    MPIR_ERR_CHKANDJUMP(seg->vaddr == (void *) -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_attach");
-
-    /* TODO: cache registered segments using AV-tree */
+
+    seg_low = MPL_ROUND_DOWN_ALIGN((uint64_t) remote_vaddr,
+                                   (uint64_t) MPIDI_XPMEM_global.sys_page_sz);
+    offset_diff = (off_t) remote_vaddr - seg_low;
+    seg_size = MPL_ROUND_UP_ALIGN(size + (size_t) offset_diff, MPIDI_XPMEM_global.sys_page_sz);
+    seg_high = seg_low + seg_size;
+
+    mpi_errno =
+        MPIDI_XPMEM_seg_avl_tree_search(MPIDI_XPMEM_global.treemaps + node_rank, seg_low, seg_high,
+                                        &seg, &voffset);
+    if (mpi_errno != MPI_SUCCESS)
```

Comment:
It is not very straightforward to me why `voffset` is needed. Can you add a comment here explaining that it is `(seg_low - seg->vaddr)` ?

Reply:
During search process, `seg` will point to the cached segment when hitting the cache. However, the attributed `vaddr` in the cached segment may not be the one we need, so `voffset` is used to tell whether there is an offset between the attached virtual address and the address we actually need.